packages/playwright/src/index.ts

```diff
@@ -523,41 +526,132 @@ function connectOptionsFromEnv() {
+    private _contentType: string,
+    private _extension: string,
+    private _doSnapshot: (page: Page, path: string) => Promise<void>) {
+  }
+
+  fixOrdinal() {
+    // Since beforeAll(s), test and afterAll(s) reuse the same TestInfo, make sure we do not
+    // overwrite previous screenshots.
+    this._ordinal = this.testInfo.attachments.filter(a => a.name === this._name).length;
+  }
+
+  private shouldCaptureUponFinish() {
+    return this._mode === 'on' ||
+        (this._mode === 'only-on-failure' && this.testInfo._isFailure()) ||
+        (this._mode === 'on-first-failure' && this.testInfo._isFailure() && this.testInfo.retry === 0);
+  }
+
+  async maybeCapture() {
+    if (!this.shouldCaptureUponFinish())
+      return;
+
+    const contexts: BrowserContext[] = [];
+    const playwright = this._artifactsRecorder._playwright;
+    for (const browserType of [playwright.chromium, playwright.firefox, playwright.webkit])
+      contexts.push(...(browserType as any)._contexts);
+
+    await Promise.all(contexts.flatMap(context => context.pages().map(page => this._snapshotPage(page, false))));
+  }
+
+  async persistTemporary() {
+    if (this.shouldCaptureUponFinish()) {
+      await Promise.all(this._temporary.map(async file => {
+        try {
+          const path = this._createAttachmentPath();
+          await fs.promises.rename(file, path);
+          this._attach(path);
+        } catch {
+        }
+      }));
+    }
+  }
+
+  async captureTemporary(context: BrowserContext) {
+    if (this._mode === 'on' || this._mode === 'only-on-failure' || (this._mode === 'on-first-failure' && this.testInfo.retry === 0))
+      await Promise.all(context.pages().map(page => this._snapshotPage(page, true)));
+  }
+
+  private _attach(screenshotPath: string) {
+    this.testInfo.attachments.push({ name: this._name, path: screenshotPath, contentType: this._contentType });
+  }
+
+  private _createAttachmentPath() {
+    const testFailed = this.testInfo._isFailure();
+    const index = this._ordinal + 1;
+    ++this._ordinal;
+    const path = this.testInfo.outputPath(`test-${testFailed ? 'failed' : 'finished'}-${index}${this._extension}`);
+    return path;
+  }
+
+  private _createTemporaryArtifact(...name: string[]) {
+    const file = path.join(this._artifactsRecorder._artifactsDir, ...name);
+    return file;
+  }
+
+  private async _snapshotPage(page: Page, temporary: boolean) {
+    if ((page as any)[this._snapshottedSymbol])
+      return;
+    (page as any)[this._snapshottedSymbol] = true;
+    try {
+      const path = temporary ? this._createTemporaryArtifact(createGuid() + this._extension) : this._createAttachmentPath();
+      await this._doSnapshot(page, path);
+      if (temporary)
+        this._temporary.push(path);
+      else
+        this._attach(path);
+    } catch {
+      // snapshot may fail, just ignore.
+    }
+  }
+
+  private get testInfo(): TestInfoImpl {
+    return this._artifactsRecorder._testInfo;
```

Comment:
ArtifactsRecorder doesn't use its _testInfo iself, consider changing fixOrdinal to make it accept ItestInfo as a parameter. Or better yet, create `SnapshotRecorder` inside `willStartTest` where `testInfo` is already available.

Reply:
ArtifactsRecorder uses `_testInfo` in `_startTraceChunkOnContextCreation`, and `SnapshotRecorder` uses `_testInfo` in more places than just `fixOrdinal`.

Reply:
What about the second part - basically confining lifetime of SnapshotRecorder to `willStartTest` - `didFinishTestFunction` ?

Reply:
`ArtifactsRecorder` already only exists for the lifetime of a test run, and `willStartTest` is called immediately after the constructor. So it wouldn't change anything about the execution, but make everything a little more complex.